### PR TITLE
CLOUDSTACK-10231: Asserted fixes for Direct Download on KVM

### PR DESCRIPTION
**Pull request by nvazquez**
Several fixes addressed:

- Dettach ISO fails when trying to detach a direct download ISO
- Fix for metalink support on SSVM agents (this closes CLOUDSTACK-10238)
- Reinstall VM from bypassed registered template (this closes CLOUDSTACK-10250)
- Fix upload certificate error message even though operation was successful
- Fix metalink download, checksum retry logic and metalink SSVM downloader

### CODE REVIEW

**Comment by rhtyd** [comment]
@nvazquez should this been seen as a blocker? /cc @DaanHoogland and others
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@rhtyd I don't think it is a blocker. It is new functionality failing, but it is nice to include if it is fixed and we need another RC.

**Comment by nvazquez** [comment]
@rhtyd agree with @DaanHoogland, it is a fix only for new Direct Download functionality

**Comment by nvazquez** [comment]
Thanks for manual verification @borisstoyanov! @DaanHoogland could you review this one as well?

File: plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java [hunk 1]
```diff
@@ -1125,11 +1106,34 @@ public Answer dettachIso(final DettachCommand cmd) {
...
         }
 
         return new Answer(cmd);
     }
 
+    /**
+     * Return data store URL from store
+     */
+    private String getDataStoreUrlFromStore(DataStoreTO store) {
+        if (!(store instanceof NfsTO) && (!(store instanceof PrimaryDataStoreTO) ||
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java]: java could really do with an unless construct here ;)

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
this closes #2411
this closes #2423
this closes #2447

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by borisstoyanov** [comment]
@blueorangutan test matrix

**Comment by nvazquez** [comment]
@borisstoyanov are we ok with test results?

**Comment by borisstoyanov** [comment]
I'm just running a smoketest against Xen7.1, to confirm these issues are there as well and we should be good @nvazquez.

**Comment by DaanHoogland** [comment]
though I only had minimum involvement I hesitate to merge myself.
Can we merge?

**Comment by DaanHoogland** [comment]
@borisstoyanov @rhtyd ^^?

**Comment by borisstoyanov** [comment]
yes @DaanHoogland, this has passed testing and it's good for merge, smoketests are showing good results and it has enough LGTMs.

**Comment by rhtyd** [comment]
@nvazquez can you squash commits based on JIRA id or you would like to squash the entire PR by a single JIRA id?
@DaanHoogland I can merge based on regression test results, I have not tested the feature manually myself. Maybe @borisstoyanov and @nvazquez can advise?

**Comment by nvazquez** [comment]
@rhtyd let me squash the entire PR to single Jira ID

**Comment by nvazquez** [comment]
@rhtyd done

**Comment by nvazquez** [comment]
@borisstoyanov @rhtyd @DaanHoogland hi guys, can we merge this one?

**Comment by rhtyd** [comment]
Merged based on tests and reviews.

**Pull request merged**